### DISCUSSION TITLE
Should guards be able to access the system?

### DISCUSSION THREAD

**Discussion by xylophonehero** [statelyai/xstate#4150] [Q&A]
I noticed that guards are unable to access the `system` like most other actors are able. I would like to read the snapshot of an actor in the system to determine how a transition in another machine behaves.

```ts
on: {
  EVENT: [
    {
      guard: ({ system }) => system.get('childActor').getSnapshot().context.myNumber > 5,
      target: 'somewhereElse',
    }
  ]
}
```

When using a choose action, I'm able to emulate this behaviour by moving the `system` check inside the action provided it is supported in the action.

```ts
// Doesn't work
on: {
  EVENT: {
    actions: choose([
      {
        guard: ({ system }) => system.get('childActor').getSnapshot().context.myNumber > 5,
        actions: raise({ type: 'NEW_EVENT' }),
      }
    ])
  }
}
// Work around (but harder to read and can't be visualised well)
on: {
  EVENT: {
    actions: raise(({ system }) => {
      if (system.get('childActor').getSnapshot().context.myNumber > 5) return { type: 'NEW_EVENT' })
      return { type: '' }
    })
  }
}
```
I could also get the actor from the context value, but this doesn't scale great when using deeply nested actors and can't be done the other way around (reading the parent snapshot from the child)

Is this something that is planned to be added or was it intentionally left out as it's a pattern that should be avoided?

**Comment by Andarist** [comment]
Ideally, guards should be pure and self-contained - relying on other actor snapshots (mutable state!) introduces a certain uncertainty into the model, its inspection capabilities etc. @davidkpiano thoughts?

>       return { type: '' }

This is something that you should never do 😅 an empty string like this is reserved in a way.

**Reply by xylophonehero** [reply]
What I'm attempting to do is stitch a whole system of actors together at the machine level without having to read the individual actor's state at the framework level and interpret how the guard works there (which is how I'm doing it right now in v4 - via provided guards in the interpret options and reading the current state of another actor I have a reference to in the framework).

It is a system of actors working together. So it feels like they should know about each other's state to know how to process certain events as a system. Especially if I know the actor will always exist in the system and if it's always kept in upto date via the sync option.

I suppose I could also model the actor system as one big state machine with nested parallel states instead of spawned actors. But using actors to split that machine up into reusable bits of logic is one of the big benefits of using xstate.

I can provide more concrete examples tomorrow if that helps. Sorry just writing my thoughts down as they come to me!

As for the empty string type, yes I will avoid using that indeed (it was just an example in this case but good to know 😁)

**Reply by davidkpiano** [reply]
> Ideally, guards should be pure and self-contained - relying on other actor snapshots (mutable state!) introduces a certain uncertainty into the model, its inspection capabilities etc. @davidkpiano thoughts?

I have mixed feelings on this, and overall I think it's _okay-ish_ to (seldom!) have impure logic in the guard.

In an ideal world, if guards were always pure, then we would be able to statically analyze the guard implementation to determine what possible combinations of `context` and `event`s can cause the guard to evaluate `true` or `false`.

_NOTE: it's not about memoizing guard return values! This is a non-goal._

However, even in that ideal world with pure guards, for non-trivial logic, you quickly run into the [Boolean satisfiability problem](https://en.wikipedia.org/wiki/Boolean_satisfiability_problem); tl;dr it becomes infeasible to do this.

So instead, for practical use-cases such as model-based testing, it is much, much easier to leave the question of which values will take each guarded transition "branch" to the _developer_. For instance, in `@xstate/test`, this is done by providing event cases, such as `{ type: 'updateName', value: '' }` and `{ type: 'updateName', value: 'validName' }`. We can still know _if_ guarded transitions are taken, which is much more important than knowing _which_ kinds values will cause each guarded transition to be taken.

When it comes to impure guards, such as:
```ts
guard: () => Math.random() < 0.5
```

the downside is that model-based testing is less reliable since you can't provide event cases that completely guarantee that every "branch" will be taken. Despite this, in practice, what matters is knowing _if_ all guards were taken. And that responsibility is up to the developer for now.

So just like there's nothing stopping a developer from writing `Math.random()` in a guard, I don't see much wrong in practice with reading from `system`; in fact, this is better than `Math.random()` because at least the developer may have control over what state an actor in the `system` can be in. It'll be less predictable, but it's a risk that the developer should be willing to take. Furthermore, it doesn't seem right to provide `system` in some parts (like actions) but not in other parts (like guards).

However, there's a more robust solution that does make the guard pure: storing the snapshot in machine `context` and using `context` in the guard.
